wormhole_attester/program/src/attestation_state.rs

```diff
@@ -14,33 +14,47 @@ use {
         AccountOwner,
         AccountState,
         Data,
-        Derive,
         Owned,
+        Peel,
+        Seeded,
     },
-    std::collections::BTreeMap,
 };
 
 /// On-chain state for a single price attestation
-#[derive(BorshSerialize, BorshDeserialize)]
+#[derive(BorshSerialize, BorshDeserialize, Default)]
 pub struct AttestationState {
     /// The last trading publish_time this attester saw
-    pub last_attested_trading_publish_time: UnixTimestamp,
+    pub last_attested_trading_publish_time: Option<UnixTimestamp>,
 }
 
-/// Top-level state gathering all known AttestationState values, keyed by price address.
-#[derive(BorshSerialize, BorshDeserialize, Default)]
-pub struct AttestationStateMap {
-    pub entries: BTreeMap<Pubkey, AttestationState>,
+impl Owned for AttestationState {
+    fn owner(&self) -> AccountOwner {
+        AccountOwner::This
+    }
 }
 
+pub struct AttestationStatePDA<'b>(
+    pub Data<'b, AttestationState, { AccountState::MaybeInitialized }>,
+);
 
-impl Owned for AttestationStateMap {
-    fn owner(&self) -> AccountOwner {
-        AccountOwner::This
+impl Seeded<&Pubkey> for AttestationStatePDA<'_> {
+    fn seeds(symbol_id: &Pubkey) -> Vec<Vec<u8>> {
+        vec![
+            "p2w-attestation-state-v1".as_bytes().to_vec(),
+            symbol_id.to_bytes().to_vec(),
+        ]
     }
 }
 
-pub type AttestationStateMapPDA<'b> = Derive<
-    Data<'b, AttestationStateMap, { AccountState::MaybeInitialized }>,
-    "p2w-attestation-state-v1",
->;
+impl<'a, 'b: 'a> Peel<'a, 'b> for AttestationStatePDA<'b> {
```

Comment:
Surprised you have to derive peel and persist manually. Why can't solitaire handle this?

Reply:
Oh, that's because automating this in a derive macro is surprisingly difficult and Solitaire cannot do this currently

Going on a bit of a framework design tangent, it's actually a very interesting topic. I think a solid dedicated `derive(Peel)` could make this go away. The thing with Solana account deserialization is that aimost any attempt at a framework will focus on adding behavior modifiers on some sort of user accounts struct, one way or another. Anchor chose to go with macro attributes as the first class citizen, where you just specify requirements in the spirit of ETH modifiers, onto struct members as part of the derive macro. Solitaire went a different direction to do as little simple codegen as necessary. The idea being to contain as much of the validation logic in regular Rust code instead of the derive macro. So the way to enforce specific rules is abstracted away from the derive macro (Notice how it says FromAccounts and not Peel). The macro only knows that all members should implement Peel and it just calls that on each account in the incoming slice on-chain. Thus most automated validation comes from stacking types like `Mut<Derive<Data<MyDataStruct, Initialized>>>`, because they implement Peel and try to cover most simple cases. There are several unanticipated problems with our approach.First,t Solana's Rust safety choices make it very hard to  come up with easy to use generics (we have to use lifetimes like 'a, 'b and so on). Also, our struct-based strategy that makes very little info available to the derive macro means that relationships between account struct members are almost impossible to express statically in struct definition

Reply:
AHHHH ok I think I get what's the problem

Reply:
You can get away with this by making : 
`pub type AttestationStatePDA<'b> =Data<'b, AttestationState, { AccountState::MaybeInitialized }>
`
instead of :
```
pub struct AttestationStatePDA<'b>(
    pub Data<'b, AttestationState, { AccountState::MaybeInitialized }>,
);
```

Reply:
I think this is just a strict simplification of the code

Reply:
That's exactly what I tried to do before I needed to add the `Seeded` impl. However, It is not possible to implement foreign traits for foreign types in Rust :( In this case, the typedef is considered equivalent to the foreign type we assign to it, so the compiler yells, because the trait is also foreign.